src/scheduler/determinism_check.rs

```diff
@@ -0,0 +1,128 @@
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+use crate::runtime::task::TaskId;
+use crate::scheduler::{Schedule, Scheduler,ScheduleStep, ScheduleRecord};
+
+/// A `DeterminismCheckScheduler` wraps an inner `Scheduler`, and when given a program,
+/// decides whether the scheduling decisions can be deterministic or not.
+/// Inner scheduler can be RoundRobin, Random, or PCT
+#[derive(Debug)]
+pub struct DeterminismCheckScheduler<S: ?Sized + Scheduler> {
+    inner: Box<S>,
+	iterations: usize,
+	inner_iterations: usize,
+	recording: bool,
+	original_schedule: Vec<ScheduleRecord>,
+	current_step: usize
+}
+
+impl<S: Scheduler> DeterminismCheckScheduler<S> {
+
+    /// Create a new `DeterminismCheckScheduler` by wrapping the given `Scheduler` implementation.
+    pub fn new(inner_iterations: usize,  inner: S) -> Self {
+        Self {
+            inner: Box::new(inner),
+			iterations: 0,
+			inner_iterations,
+			original_schedule: Vec::new(),
+			recording: false,
+			current_step: 0
+        }
+    }
+}
+
+
+impl<S: Scheduler> Scheduler for DeterminismCheckScheduler<S> {
+    fn new_execution(&mut self) -> Option<Schedule> {
+
+		if self.iterations % self.inner_iterations == 0 {
+			// Start a new recording
+			self.recording = true;
+			self.original_schedule.clear();
+		} else {
+			// Create a new execution to test against prior recording
+			
+			if self.recording {
+				self.recording = false;
+			}
+
+			if self.current_step != self.original_schedule.len() {
+				panic!("Current execution ended earlier than original execution\n\tOriginal length: {}, acutal length: {}",
+																self.original_schedule.len(), self.current_step);
```

Comment:
I think our style is lowercase, and to not include newlines in panics, so:
```suggestion
				panic!("possible nondeterminism: current execution ended earlier than expected (expected length {} but ended after {})", self.original_schedule.len(), self.current_step);
```